File: mobile/src/main/java/com/alexstyl/specialdates/events/namedays/calendar/resource/RomanianSpecialNamedays.java
package com.alexstyl.specialdates.events.namedays.calendar.resource;

import com.alexstyl.specialdates.date.Date;
import com.alexstyl.specialdates.events.namedays.NameCelebrations;
import com.alexstyl.specialdates.events.namedays.NamesInADate;
import com.alexstyl.specialdates.events.namedays.calendar.EasternNameday;
import com.alexstyl.specialdates.events.namedays.calendar.EasternNamedaysExtractor;

import java.util.ArrayList;
import java.util.List;

final class RomanianSpecialNamedays implements SpecialNamedays {

    private final RomanianNamedays namedays;

    public static SpecialNamedays from(NamedayJSON namedayJSON) {
        EasternNamedaysExtractor extractor = new EasternNamedaysExtractor(namedayJSON.getSpecial());
        List<EasternNameday> easternNamedays = extractor.parse();
        List<String> names = namesOf(easternNamedays);
        RomanianNamedays namedays = RomanianNamedays.from(names);
        return new RomanianSpecialNamedays(namedays);
    }

    private static List<String> namesOf(List<EasternNameday> easternNamedays) {
        List<String> names = new ArrayList<>();
        for (EasternNameday easternNameday : easternNamedays) {
            names.addAll(easternNameday.getNamesCelebrating());
        }
        return names;
    }

    public RomanianSpecialNamedays(RomanianNamedays namedays) {
        this.namedays = namedays;
    }

    @Override
    public List<String> getAllNames() {
        return namedays.getAllNames();
    }

    @Override
    public NamesInADate getNamedayOn(Date date) {
        return namedays.getNamedaysFor(date);
    }

    @Override
    public NameCelebrations getNamedaysFor(String name, int year) {
        return namedays.getNamedaysFor(name, year);
    }
}

File: memento/src/main/java/com/alexstyl/specialdates/events/namedays/calendar/resource/RomanianSpecialNamedays.kt
package com.alexstyl.specialdates.events.namedays.calendar.resource

import com.alexstyl.specialdates.date.Date
import com.alexstyl.specialdates.events.namedays.NameCelebrations
import com.alexstyl.specialdates.events.namedays.NamesInADate
import com.alexstyl.specialdates.events.namedays.calendar.EasternNameday
import com.alexstyl.specialdates.events.namedays.calendar.EasternNamedaysExtractor


internal class RomanianSpecialNamedays private constructor(private val namedays: RomanianNamedays) : SpecialNamedays {

    override val allNames: List<String>
        get() = namedays.allNames

    override fun getNamedayOn(date: Date): NamesInADate {
        return namedays.getNamedaysFor(date)
    }

    override fun getNamedaysFor(name: String, year: Int): NameCelebrations {
        return namedays.getNamedaysFor(name, year)
    }

    companion object {

        fun from(namedayJSON: NamedayJSON, romanianEasterCalculator: RomanianEasterSpecialCalculator): SpecialNamedays {
            val extractor = EasternNamedaysExtractor(namedayJSON.special)
            val easternNamedays = extractor.parse()
            val names = namesOf(easternNamedays)
            val namedays = RomanianNamedays(romanianEasterCalculator, names)
            return RomanianSpecialNamedays(namedays)
        }

        private fun namesOf(easternNamedays: List<EasternNameday>): ArrayList<String> {
            val names = ArrayList<String>()
            for (easternNameday in easternNamedays) {
                names.addAll(easternNameday.namesCelebrating)
            }
            return names
        }
    }
}
